### ISSUE TITLE
Pizzly-js not supported on Node.js (only client-side web)

### ISSUE THREAD

**Issue by nabilfreeman** [Bearer/Pizzly#66]
PIzzly-js is built with es6 syntax which is fully supported on Node 13 and above.

However on Node 8-12 the only way to enable this is by passing `--experimental-modules` to the Node executable.

In some situations this isn't possible (for example applications running on AWS Lambda which at time of writing doesn't support versions above 12 for Node.js), so some additional steps are needed for compatibility.

Ideas to transpile `pizzly-js`:
- Upgrade Node 😛
- postinstall hook to transpile `pizzly-js`
- Deployment hook to transpile `pizzly-js`

I don't think this is really in the scope of the project itself but I think it would be a good idea to leave a note on the `pizzly-js` npm page with minimum required node version and/or pipeline adjustments that might be needed.

https://www.npmjs.com/package/pizzly-js

Thinking about it more, I think this project _could_ potentially implement this via a postinstall hook on the `pizzly-js` package.json that detects node version and acts accordingly.

Happy to give this a try if it's something you want to support.

**Comment by nabilfreeman** [comment]
Ok, I had to jump through a lot of hoops to get this working on AWS Lambda and Node.js v12.

Here is what went into my postinstall script:

```
# lord forgive me for my sins

# delete es6 index.js that is referenced by `main` in package.json
rm -rf ./node_modules/pizzly-js/dist/index.js
# copy the browser umd file to index.js so it will be the one required
cp ./node_modules/pizzly-js/dist/index.umd.js ./node_modules/pizzly-js/dist/index.js
# URL is global in browsers, here we need it from the url node module, sed prepends to the top of file.
sed -i.bak '1s/^/const { URL } = require("url");/' ./node_modules/pizzly-js/dist/index.js
# `node-fetch` npm module is required to polyfill fetch into Node.js
sed -i.bak '1s/^/global.fetch = require("node-fetch");/' ./node_modules/pizzly-js/dist/index.js
# window doesn't exist in Node, connect will be completely broken now but I don't need it
sed -i.bak 's/window/global/' ./node_modules/pizzly-js/dist/index.js
```

And it works!!!

Now I think it's clear that `pizzly-js` is actually just meant for browser use which explains the es6 development style mentioned above.

However it is possible to use this library on Node in the same way as Bearer. Are there any plans to have official Node SDK support? Very happy to help out on this so I can remove the above code from postinstall!

**Comment by markmichon** [comment]
Hi @nabilfreeman, as you mentioned the client is intended for browsers (the reliance on fetch is the big blocker, more than es6, as well as the back and forth between windows to perform the OAuth dance), but I'll let @Frenchcooc chime in with any details on any future plans for a serverside implementation. 

For a less involved workaround, you *should* be able to perform proxying by hitting the `/proxy/INTEGRATION_NAME/endpoint` endpoint, and passing the `Pizzly-Auth-Id` header as shown in the curl example.

Using this approach, any node request library will suffice (like node-fetch). For example:

```js
const fetch = require('node-fetch')

fetch('https://your-pizzly-auth-server.com/proxy/github/user', { headers: { 'Pizzly-Auth-Id': 'YOUR_AUTH_ID' } })
  .then(res => { ... })
```

**Comment by nabilfreeman** [comment]
Hey @markmichon, thanks for the speedy reply. Yes you're right that should work well, bonus points for using node-fetch so on topic!! Haha. Just need to add basic auth to that for secured hosts.

For us we had a lot of implementations built around the original `@bearer/node` client which had the same syntax, so this method was easier for getting the migration done in time for the shutdown date.

At a later date based on plans for an SDK I'll clean things up and go either way. I think you could easily repackage some of the classes you've written into both a web SDK and node SDK that build from the same codebase though.

Lastly on the `/proxy` raw request I did manage to figure that out but found the documentation of it a bit lacking so had to look at the auth server module to figure out what it was expecting? I opened an issue here about it: #65

**Comment by nabilfreeman** [comment]
Update: I ran into some more issues with the Pizzly JS library using public key instead of secret key so I quickly wrote my own module today for using Pizzly on back-end.

It depends on Axios and requires 

Gist here:
https://gist.github.com/nabilfreeman/5ea3ab66075a0caa8ecb1347459b8f08

Use like this:

```
const Pizzly = require('./pizzly');

const { data: lists } = await Pizzly(
	'trello',
	'get',
	`/boards/${this.BOARD_ID}/lists`,
);
```

**Comment by Frenchcooc** [comment]
Hi Nabil. Thanks a lot for sharing your Gist. As said by Mark, `pizzly-js` works only within the browser. Not having a Node client is really a pain, I totally agree. I'll take some time this week to release an official SDK for Node. I feel it's really missing.

Trying to run `pizzly-js` from a node runtime should have triggered an error though. Did you get over it? I might improve the error message here!

**Comment by nabilfreeman** [comment]
There wasn't any error @Frenchcooc, instead I got several layers of errors mainly about ES6 syntax (can't run Node 13 on Lambda for native ES6 support). So any Node version 12 or below wouldn't be able to even import the `pizzly-js` library.

For that reason I hacked it to use the compiled index.umd.js module via a postinstall script above. It almost worked, but when doing `get` requests `pizzly-js` was automatically specifying the auth as public key when I was supplying a secret key.

I actually think I could have just placed the public key in environment and then everything would have worked, but I decided it would be an anti-pattern to use public key on private server!
